### PR TITLE
[video_player] Add web implementation using platform interface

### PR DESCRIPTION
**Pull request by cbenhagen**
## Description

Adding video_player_web implementation using the platform interface from https://github.com/flutter/plugins/pull/2273

Fixes https://github.com/flutter/flutter/issues/45290

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Comment by cbenhagen** [comment]
cc @hterkelsen

**Comment by cbenhagen** [comment]
@hterkelsen this should be ready as well. Thanks for all your help!

**Comment by cbenhagen** [comment]
Added some basic tests and made the VideoPlayer class private.

On the web we can't just autoplay videos with sound (See https://goo.gl/xX8pDD). How should the plugin handle this? In the tests I am muting the video before playing. Is it ok to have this throw a runtime exception? How can we best help users to avoid that?

**Comment by cbenhagen** [comment]
@hterkelsen PTAL

(Not sure if you have already seen this PR and are subscribed. If you are, please excuse the noise.)

File: packages/video_player/video_player_web/lib/video_player_web.dart [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+
+  @override
+  Future<void> dispose(int textureId) async {
+    _videoPlayers[textureId].dispose();
+    _videoPlayers.remove(textureId);
+    return null;
+  }
+
+  void _disposeAllPlayers() {
+    _videoPlayers
```
Comment (hterkelsen) [inline packages/video_player/video_player_web/lib/video_player_web.dart]: nit: maybe `_videoPlayers.keys.forEach` would be a little nicer?

File: packages/video_player/video_player_web/lib/video_player_web.dart [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+  }
+
+  @override
+  Stream<VideoEvent> videoEventsFor(int textureId) {
+    return _videoPlayers[textureId].eventController.stream;
+  }
+
+  @override
+  Widget buildView(int textureId) {
+    return HtmlElementView(viewType: textureId.toString());
```
Comment (hterkelsen) [inline packages/video_player/video_player_web/lib/video_player_web.dart]: Let's make the viewType something like `videoPlayer-$textureId`

File: packages/video_player/video_player_web/lib/video_player_web.dart [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+  void setupVideoPlayer() {
+    videoElement = VideoElement()
+      ..src = uri.toString()
+      ..autoplay = false
+      ..controls = false
+      ..style.border = 'none';
+
+    // ignore: undefined_prefixed_name
+    ui.platformViewRegistry.registerViewFactory(
```
Comment (hterkelsen) [inline packages/video_player/video_player_web/lib/video_player_web.dart]: We should probably add initialization parameters to HTML platform views so you don't need to register a new view type for videos that are the same except for the `src` parameter. That shouldn't be a prerequisite for landing this, though.

Can you add a TODO to use initialization parameters once they are available?

File: packages/video_player/video_player_web/pubspec.yaml [hunk 4]
```diff
@@ -0,0 +1,30 @@
+name: video_player_web
+description: Web platform implementation of video_player
+author: Flutter Team <flutter-dev@googlegroups.com>
+homepage: https://github.com/flutter/plugins/tree/master/packages/video_player/video_player_web
+version: 0.0.1
```
Comment (hterkelsen) [inline packages/video_player/video_player_web/pubspec.yaml]: Prefer `0.1.0` so this package gets a slightly better score on pub

File: packages/video_player/video_player_web/lib/video_player_web.dart (hunk 1 above)
Comment (cbenhagen) [reply packages/video_player/video_player_web/lib/video_player_web.dart]: Much nicer indeed.

File: packages/video_player/video_player_web/lib/video_player_web.dart (hunk 2 above)
Comment (cbenhagen) [reply packages/video_player/video_player_web/lib/video_player_web.dart]: Done

File: packages/video_player/video_player_web/lib/video_player_web.dart (hunk 3 above)
Comment (cbenhagen) [reply packages/video_player/video_player_web/lib/video_player_web.dart]: Not sure how that would look like yet. Added a TODO with your name. Hope that's ok.

File: packages/video_player/video_player_web/pubspec.yaml (hunk 4 above)
Comment (cbenhagen) [reply packages/video_player/video_player_web/pubspec.yaml]: Done

File: packages/video_player/video_player_web/lib/video_player_web.dart [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+  }
+
+  @override
+  Stream<VideoEvent> videoEventsFor(int textureId) {
+    return _videoPlayers[textureId].eventController.stream;
+  }
+
+  @override
+  Widget buildView(int textureId) {
+    return HtmlElementView(viewType: textureId.toString());
```
Comment (hterkelsen) [inline packages/video_player/video_player_web/lib/video_player_web.dart]: You use viewType `textureId.toString()` here but registered at `videoPlayer-$textureId` below.

Comment (cbenhagen) [reply packages/video_player/video_player_web/lib/video_player_web.dart]: Fixed

**Comment by cbenhagen** [comment]
@hterkelsen what are your thoughts on the autoplay issue? Is this something the plugin should try to handle? If yes, do we need to improve it now or can we do that at a later stage?

**Comment by hterkelsen** [comment]
We should document the autoplay issue issue with a link to the Chrome Autoplay Policy changes article. I don't think we should block landing this plugin on it, though. As long as the `VideoPlayerController` can play and pause videos with sound on user taps, then we should be good.

File: packages/video_player/video_player_web/lib/video_player_web.dart [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+
+  final StreamController<VideoEvent> eventController =
+      StreamController<VideoEvent>();
+
+  final Uri uri;
+  final int textureId;
+  VideoElement videoElement;
+  bool isInitialized = false;
+
+  void setupVideoPlayer() {
```
Comment (collinjackson) [inline packages/video_player/video_player_web/lib/video_player_web.dart]: Nit: I would probably name this method something like `initialize`. The class is already named `_VideoPlayer` and it's redundant to put the class name `VideoPlayer` in the method name.

Comment (cbenhagen) [reply packages/video_player/video_player_web/lib/video_player_web.dart]: Fixed

**Comment by cbenhagen** [comment]
Please do not merge yet. I just found a bug I'd like to fix first.

**Comment by cbenhagen** [comment]
Done. Sorry for the noise and the missed bug. This should be ready to merge now. @hterkelsen I also added a note about the potential autoplay issues to the readme.

**Comment by hterkelsen** [comment]
I'll land this as soon as the tests go green. @cbenhagen Thanks so much for this contribution, and for being so patient with all the reviews! This is a huge improvement for all Flutter for web users!

**Pull request merged**